File: the-odin-project/javascript/largest-prime.js
// The prime factors of 13195 are 5, 7, 13, and 29
// What is the largest prime factor of the number 600851475143?

var isFactor = function(num1, num2) {
	return num1 % num2 === 0;
}

var contains = function(list, obj) {
	var k = list.length;
	while (k--) {
		if (list[k] === obj) {
			return true;
		}
	}
	return false;
}

var listFactors = function(multiple) {
	var factors = [];
	for (i = 1; i < multiple/2+1; i++) {
		if (contains(factors, i)) {
			break;
		}
		if (isFactor(multiple, i)) {
			factors.push(i);
			if (!contains(factors, multiple/i)) {
				factors.push(multiple/i);
			}
		}
	}
	return factors;
}

var isPrime = function(number) {
	var factors = listFactors(number);
	return factors.length <= 2;
}

var listPrimeFactors = function(integer) {
	var factors = listFactors(integer);
	var primes = [];
	for (j = 0; j < factors.length; j++) {
		if (isPrime(factors[j])) {
			primes.push(factors[j])
			continue;
		}
	}
	return primes;
}

var largestPrimeFactor = function(int) {
	var primeFactors = listPrimeFactors(int);
	return primeFactors.pop();
}

largestPrimeFactor(600851475143);

File: ruby/ruby_101/largest_prime.rb
# The prime factors of 13195 are 5, 7, 13, and 29
# What is the largest prime factor of the number 600851475143?

def is_factor multiple, factor
  multiple % factor == 0
end

def list_factors number
  factors = []
  (1..number).each do |factor|
    break if factors.include?(factor)
    factors << factor << (number/factor) if is_factor(number, factor)
  end
  factors.sort.uniq
end

def is_prime number
  list_factors(number).length <= 2
end

def list_prime_factors number
  list_factors(number).select { |i| is_prime(i) }
end

def largest_prime_factor number
  list_prime_factors(number).max
end

puts "The largest prime factor of 600851475143 is :#{largest_prime_factor(600851475143)}"